### PR TITLE
Add tsh e2e tests with various security features enabled

### LINKED ISSUE #24094: Improve test coverage for Teleport when using per-session MFA and device trust
Improve test coverage for Teleport when using per-session MFA, device trust, headless SSO, and agent forwarding with various Teleport security features enabled.

Ensure all of the following combinations have automated test coverage.

|                 | `tsh ssh nodename` | `tsh ssh FOO=BAR` | `tsh ssh public_addr` | `tsh scp` upload | `tsh scp` download |
|-----------------|--------------------|-------------------|-----------------------|------------------|--------------------|
| max sessions    |                    |                   |                       |                  |                    |
| max connections |                    |                   |                       |                  |                    |

### PR DESCRIPTION
**Pull request by Joerger**
To properly test device trust, the test had to be put in `teleport/e`. This required:
 - Separating the tsh library from its `main` package.
 - Providing `tool/teleport/mock/mock.go` to provide a similar mocked teleport to what's used in current `tsh`, `tctl`, and other tests (integration, etc.). This common mock can be used to replace the individual mocks in the aforementioned tests, but that would be a separate PR.

Needed for this e PR - https://github.com/gravitational/teleport.e/pull/1488

Based off of https://github.com/gravitational/teleport/pull/27039

Closes https://github.com/gravitational/teleport/issues/24094

### CODE REVIEW

File: lib/client/session.go [hunk 1]
```diff
@@ -270,7 +270,7 @@ func selectKeyAgent(tc *TeleportClient) agent.ExtendedAgent {
 	switch tc.ForwardAgent {
 	case ForwardAgentYes:
 		log.Debugf("Selecting system key agent.")
-		return tc.localAgent.systemAgent
```
Comment (tobiaszheller) [inline lib/client/session.go:273]: i guess base of PR should be `joerger/fix-agent-forwarding-without-local-agent-keys`

**Review by tobiaszheller — COMMENTED** [review]
looks mostly good, left few comments.

**Comment by tobiaszheller** [comment]
@Joerger please also look at flaky test detector

File: tool/teleport/mock/mock.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+						break
+					}
+				}
+			}
+
+			if registered == len(dbs) {
+				return
+			}
+		case <-ctx.Done():
+			t.Fatal("databases not registered after 10s")
```
Comment (codingllama) [inline tool/teleport/mock/mock.go:196]: nit:

```suggestion
			t.Fatal("Databases not registered after 10s")
```

File: tool/teleport/mock/mock.go [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+		}
+		cfg.SSH.Labels[key] = value
+	})
+}
+
+type cliModules struct{}
+
+// BuildType returns build type (OSS or Enterprise)
+func (p *cliModules) BuildType() string {
+	return "CLI"
```
Comment (codingllama) [inline tool/teleport/mock/mock.go:262]: Comment says "OSS" or "Enterprise", this returns "CLI". Which is correct?

File: tool/teleport/mock/mock.go [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+}
+
+// PrintVersion prints the Teleport version.
+func (p *cliModules) PrintVersion() {
+	fmt.Printf("Teleport CLI\n")
```
Comment (codingllama) [inline tool/teleport/mock/mock.go:267]:
```suggestion
	fmt.Println("Teleport CLI")
```

File: tool/teleport/mock/mock.go
[the remark this answers is not in the document]
Comment (codingllama) [reply tool/teleport/mock/mock.go:1]: +1

**Comment by codingllama** [comment]
Thanks for tackling this, Brian!

File: lib/devicetrust/authn/export.go
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/devicetrust/authn/export.go:1]: Done - https://github.com/gravitational/teleport/pull/27039, I'll base this PR off that one for now.

**Comment by Joerger** [comment]
> @Joerger please also look at flaky test detector

The test in https://github.com/gravitational/teleport.e/pull/1488 is pretty long, so it will probably need to be added as an edge case to pass the "slow test detector" (times out from tests longer than 6 seconds)

File: tool/teleport/mock/mock.go [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+func init() {
+	// If the test is re-executing itself, execute the command that comes over
+	// the pipe. Used to test tsh ssh and tsh scp commands.
+	if srv.IsReexec() {
+		common.Run(common.Options{Args: os.Args[1:]})
+		return
+	}
+
+	var err error
+	ports, err = utils.GetFreeTCPPorts(5000, utils.PortStartingNumber)
```
[the remark this answers is not in the document]
Comment (codingllama) [reply tool/teleport/mock/mock.go]: That constant is the default for the offset, by the looks of it.

[the remark this answers is not in the document]
Comment (codingllama) [reply tool/teleport/mock/mock.go]: That seems much more reasonable, thanks.

File: lib/devicetrust/authn/export.go
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/devicetrust/authn/export.go:1]: Thanks for that!

File: tool/teleport/testing/test_server.go [hunk 6]
```diff
@@ -14,7 +14,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package mock
+// Package testing provides functions for creating test servers for testing.
+package testing
```
Comment (codingllama) [inline tool/teleport/testing/test_server.go:18]: This is a nasty name conflict with the standard "testing" package.

I suggest a different variant like `testenv` (which I do have in a few places) or `testonly` (even if you end up using an import alias in the end).

File: tool/tsh/common/tsh.go [hunk 7]
```diff
@@ -450,6 +451,11 @@ type CLIConf struct {
 
 	// HeadlessAuthenticationID is the ID of a headless authentication.
 	HeadlessAuthenticationID string
+
+	// dtAuthnCeremony allows tests to override the default device
```
Comment (codingllama) [inline tool/tsh/common/tsh.go:455]:
```suggestion
	// DTAuthnCeremony allows tests to override the default device
```

File: lib/client/api.go [hunk 8]
```diff
@@ -430,6 +430,21 @@ type Config struct {
...
+
+	// DTAuthnCeremony allows tests to override the default device
+	// authentication function.
+	// Defaults to [dtauthn.NewCeremony].
+	DTAuthnCeremony dtauthn.CeremonyI
+
+	// dtAutoEnroll allows tests to override the default device auto-enroll
+	// function.
+	// Defaults to [dtenroll.AutoEnroll].
+	dtAutoEnroll dtAutoEnrollFunc
```
Comment (codingllama) [inline lib/client/api.go:447]: nits: publics before privates?

```suggestion
	// DTAuthnCeremony allows tests to override the default device
	// authentication function.
	// Defaults to [dtauthn.NewCeremony].
	DTAuthnCeremony dtauthn.CeremonyI

	// dtAttemptLoginIgnorePing and dtAutoEnrollIgnorePing allow Device Trust
	// tests to ignore Ping responses.
	// Useful to force flows that only typically happen on Teleport Enterprise.
	dtAttemptLoginIgnorePing, dtAutoEnrollIgnorePing bool

	// dtAutoEnroll allows tests to override the default device auto-enroll
	// function.
	// Defaults to [dtenroll.AutoEnroll].
	dtAutoEnroll dtAutoEnrollFunc
```

File: tool/teleport/testing/test_server.go (hunk 6 above)
Comment (strideynet) [reply tool/teleport/testing/test_server.go:18]: +1 for testenv

**Review by tobiaszheller — APPROVED** [review]
lgtm, just apply Alan comment

**Review by codingllama — APPROVED** [review]

**Comment by Joerger** [comment]
@r0mant is there a way to bypass the flaky test detector for this PR? It looks like it is trying to run all of the tests in `tool/tsh/common`, which was moved from `tool/tsh` in this PR:

```
2023-05-31T17:49:13.0456936Z CGO_ENABLED=1 CGO_LDFLAGS="-Wl,-Bstatic  -Wl,-Bdynamic -Wl,--as-needed" go test -cover -json -tags "pam  libfido2  piv"  -run "^TestAccessRequestSearch$|^Test_expandAliasDefinition$|^Test_findAliasCommand$|^Test_getSeenAliases$|^Test_markAliasSeen$|^Test_getAliasDefinition$|^Test_runAliasCommand$|^TestAWS$|^TestAzure$|^Test_getAzureIdentityFromFlags$|^Test_getMSISecret$|^Test_formatAzureIdentities$|^Test_filterApps$|^Test_getGCPServiceAccountFromFlags$|^Test_formatGCPServiceAccounts$|^Test_gcpApp_Config$|^TestFormatAppConfig$|^TestWriteSSHConfig$|^TestDatabaseLogin$|^TestLocalProxyRequirement$|^TestListDatabase$|^TestFormatDatabaseListCommand$|^TestFormatConfigCommand$|^TestDBInfoHasChanged$|^TestFormatDatabaseConnectArgs$|^TestGetDefaultDBNameAndUser$|^TestKube$|^Test_maybeStartKubeLocalProxy$|^Test_isKubectlConfigCommand$|^Test_extractKubeConfigAndContext$|^Test_overwriteKubeconfigFlagInArgs$|^Test_overwriteKubeconfigFlagInEnv$|^TestOptions$|^TestSSH$|^TestSSHLoadAllCAs$|^TestProxySSH$|^TestProxySSHJumpHost$|^TestTSHProxyTemplate$|^TestTSHConfigConnectWithOpenSSHClient$|^TestEnvVarCommand$|^TestList$|^TestFormatCommand$|^Test_chooseProxyCommandTemplate$|^TestPrintProxyAWSTemplate$|^TestCheckProxyAWSFormatCompatibility$|^TestWriteMovieCanBeCanceled$|^TestWriteMovieDoesNotSupportSSH$|^TestWriteMovieMultipleScreenSpecs$|^TestWriteMovieWritesOneFrame$|^TestWriteMovieWritesManyFrames$|^TestResolveDefaultAddr$|^TestResolveDefaultAddrNoCandidates$|^TestResolveDefaultAddrSingleCandidate$|^TestResolveDefaultAddrTimeout$|^TestResolveNonOKResponseIsAnError$|^TestResolveUndeliveredBodyDoesNotBlockForever$|^TestResolveDefaultAddrTimeoutBeforeAllRacersLaunched$|^TestResolveDefaultAddrHTTPProxy$|^TestLoadConfigFromProfile$|^TestRemoteTctlWithProfile$|^TestSetAuthServerFlagWhileLoggedIn$|^TestMain$|^TestAlias$|^TestFailedLogin$|^TestOIDCLogin$|^TestLoginIdentityOut$|^TestRelogin$|^TestSwitchingProxies$|^TestMakeClient$|^TestSSHOnMultipleNodes$|^TestSSHAccessRequest$|^TestAccessRequestOnLeaf$|^TestSSHHeadless$|^TestFormatConnectCommand$|^TestEnvFlags$|^TestKubeConfigUpdate$|^TestSetX11Config$|^TestAuthClientFromTSHProfile$|^TestSerializeVersion$|^TestSerializeApps$|^TestSerializeAppsEmpty$|^TestSerializeAppConfig$|^TestSerializeDatabases$|^TestSerializeDatabasesEmpty$|^TestSerializeDatabaseEnvironment$|^TestSerializeDatabaseConfig$|^TestSerializeNodes$|^TestSerializeNodesEmpty$|^TestSerializeClusters$|^TestSerializeProfiles$|^TestSerializeProfilesNoOthers$|^TestSerializeProfilesNoActive$|^TestSerializeProfilesWithEnvVars$|^TestSerializeEnvironment$|^TestSerializeAccessRequests$|^TestSerializeKubeSessions$|^TestSerializeKubeClusters$|^TestSerializeMFADevices$|^TestListDatabasesWithUsers$|^TestForwardingTraces$|^TestExportingTraces$|^TestShowSessions$|^TestMakeProfileInfo_NoInternalLogins$|^TestLoadConfigNonExistingFile$|^TestLoadConfigEmptyFile$|^TestLoadAllConfigs$|^TestTshConfigMerge$|^TestProxyTemplatesApply$|^TestProxyTemplatesMakeClient$" ./tool/tsh/common -race -shuffle on -count 100 \
```

**Pull request merged**